kaggle_environments/envs/lux_ai_2022/luxai2022/actions.py

```diff
@@ -0,0 +1,331 @@
+    elif a_type == 2:
+        return PickupAction(a[2], a[3], repeat=a[4])
+    elif a_type == 3:
+        return DigAction(repeat=a[4])
+    elif a_type == 4:
+        return SelfDestructAction(repeat=a[4])
+    elif a_type == 5:
+        return RechargeAction(a[3], repeat=a[4])
+    else:
+        raise ValueError(f"Action {a} is invalid type, {a[0]} is not valid")
+
+
+# a[1] = direction (0 = center, 1 = up, 2 = right, 3 = down, 4 = left)
+move_deltas = np.array([[0, 0], [0, -1], [1, 0], [0, 1], [-1, 0]])
+
+
+def validate_actions(env_cfg: EnvConfig, state: 'State', actions_by_type, weather_cfg, verbose=1):
+    """
+    validates actions and logs warnings for any invalid actions. Invalid actions are subsequently not evaluated
+    """
+    actions_by_type_validated = defaultdict(list)
+    valid_action = True
+
+    def invalidate_action(msg):
+        nonlocal valid_action
+        valid_action = False
+        if verbose > 0: print(f"{state.real_env_steps}: {msg}")
+
+    for unit, transfer_action in actions_by_type["transfer"]:
+        valid_action = True
+        unit: luxai_unit.Unit
+        transfer_action: TransferAction
+        if transfer_action.resource > 4 or transfer_action.resource < 0:
+            invalidate_action(
+                f"Invalid Transfer Action for unit {unit}, transferring invalid resource id {transfer_action.resource}"
+            )
+            continue
+        # if transfer_action.transfer_amount < 0: do not need to check as action space permits range of [0, max_transfer_amount] anyway
+        # TODO - check what happens if transferring with direction center?
+        resource_id = transfer_action.resource
+        amount = transfer_action.transfer_amount
+        if resource_id == 0:
+            if unit.cargo.ice < amount:
+                invalidate_action(
+                    f"Invalid Transfer Action for unit {unit} - Tried to transfer {amount} ice but only had {unit.cargo.ice}"
+                )
+                continue
+        elif resource_id == 1:
+            if unit.cargo.ore < amount:
+                invalidate_action(
+                    f"Invalid Transfer Action for unit {unit} - Tried to transfer {amount} ore but only had {unit.cargo.ore}"
+                )
+                continue
+        elif resource_id == 2:
+            if unit.cargo.water < amount:
+                invalidate_action(
+                    f"Invalid Transfer Action for unit {unit} - Tried to transfer {amount} water but only had {unit.cargo.water}"
+                )
+                continue
+        elif resource_id == 3:
+            if unit.cargo.metal < amount:
+                invalidate_action(
+                    f"Invalid Transfer Action for unit {unit} - Tried to transfer {amount} metal but only had {unit.cargo.metal}"
+                )
+                continue
+        elif resource_id == 4:
+            if unit.power < amount:
+                invalidate_action(
```

Comment:
needs to have enough power for the transfer + the transfer action cost?

Reply:
or did we remove transfer cost?

Reply:
no transfer costs